### PR TITLE
Remove "parent" and "slug" fields

### PR DESCRIPTION
**Pull request by meshy**
TODO:

- [x] Remove `parent` field.
- [x] Remove `slug` field.
- [x] Add validators to `url` field.
 - Starts in `/`.
 - Ends in `/`.
 - Cannot contain `?` (to avoid GET args)
 - [x] **Are there any other invalid chars?** `..` perhaps?
- [x] Remove logic for cascading URL renames from `Route.save()`

**Deferred to #80 and rebased away from this PR**
- Add `Route.objects.swap_routes(first, second)` method.
- Add `Route.objects.swap_branches(first, second)` method.
- Add `Route.objects.move_branch(old_url, new_url)`.
- Add `Route().get_parent()`
  - Should this be a manager method?
  - How about a cached property, `parent`?
- Add `Route().get_ancestors()`
  - Should this be a manager method?
  - How about a cached property, `ancestors`?

**Deferred to #79**
- Add custom admin view for cascading URL moves.
- Make `url` a first-class (not generated) field in the admin.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 ### Backwards incompatible
 
-* Removed dependancy upon `django-polymorphic-tree`.
+* Removed dependency upon `django-polymorphic-tree`.
 * Destroyed and recreated migrations.
```
Comment (meshy) [inline CHANGELOG.md]: Needs more here.

**Comment by meshy** [comment]
@Ian-Foote, any comments on this? In particular, I'm interested on feedback on how I've done the validators, and if you think I should add more? Things that spring to mind are things like consecutive `//`, or `..`, or anything else that might make the `url` an invalid "path" url fragment. What do you think?

**Comment by Ian-Foote** [comment]
Your validators seem fine to me. Is there a downside to adding more?

**Comment by meshy** [comment]
> Your validators seem fine to me.

Thanks `:)`

> Is there a downside to adding more?

Not that I'm aware of. I'm pretty sure that the "no `//`" validator is worthwhile. Not sure about `..`. Are there any more that you can think of?

**Comment by Ian-Foote** [comment]
Do you want to handle https://perishablepress.com/stop-using-unsafe-characters-in-urls/ at this level? Your explicit handling of `?` suggests to me that you do.

**Comment by meshy** [comment]
> Do you want to handle reserved and unsafe characters...?

Possibly. I'm not sure of the consequences. Will that stop the use of unicode in URLs? Is that really desirable?

> Your explicit handling of `?` suggests to me that you do.

The reason I added that one is because django strips GET args out of urls. Anything from a `?` onwards could not possibly be used for routing, so I thought it best to remove it.

**Comment by meshy** [comment]
> django strips GET args out of urls

Further to that (and thanks to your link), I think I should be disallowing `#` as well at least.

**Comment by Ian-Foote** [comment]
I think internationalised urls are a separate issue.

**Comment by meshy** [comment]
Ok, so I've added a few more.
- Subpaths cannot contain only `.`. This prevents parent directory paths `/../` etc.
- Paths cannot contain `#`. This prevents fragments from sneaking in.
- Paths cannot contain `//`. This prevents `//` at the beginning to make a protocol, and subpaths with zero-length names.

**Comment by meshy** [comment]
I'm not sure that this is the correct stage to be dealing with "unsafe" chars. That should probably happen when the url is output to a template, with `{{ route.url|urlencode }}`. @Ian-Foote @KevinEtienne do you agree with that? If so, happy to see merged?

**Comment by meshy** [comment]
Ok, after discussion IRL, this looks good to go. Merging.

**Pull request merged**